### PR TITLE
feat: compute op1 and deduce op0

### PR DESCRIPTION
**Pull request by greged93**
Updates the error handling and implements the compute op1 and deduce op0.

### CODE REVIEW

File: src/result/error.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  VMError: 'VMError',
+  InstructionError: 'InstructionError',
+  RunContextError: 'RunContextError',
+} as const;
+
+export class BaseError extends Error {
+  type: string;
+  message: string;
+
+  constructor(type: string, message: string) {
```
Comment (Eikix) [inline src/result/error.ts:16]:
```suggestion
  constructor(type: typeof ErrorType, message: string) {
```

File: src/result/error.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  Int16Error: 'Int16Error',
+  UintError: 'UintError',
+  MemoryError: 'MemoryError',
+  VMError: 'VMError',
+  InstructionError: 'InstructionError',
+  RunContextError: 'RunContextError',
+} as const;
+
+export class BaseError extends Error {
+  type: string;
```
Comment (Eikix) [inline src/result/error.ts:13]:
```suggestion
  type: typeof ErrorType;
```

File: src/vm/virtualMachine.ts [hunk 3]
```diff
@@ -28,20 +25,113 @@ export class VirtualMachine {
...
+
     // return this.runInstruction();
     throw new Error('TODO: Not Implemented');
   }
 
-  computeOperands(instruction: Instruction) {}
+  // Compute the operands of an instruction. The VM can either
+  // fetch them from memory and if it fails to do so, it can
+  // deduce them from the instruction itself.
+  computeOperands(instruction: Instruction): Result<void> {
```
Comment (Eikix) [inline src/vm/virtualMachine.ts:52]: why does it return void?
is it a todo?

File: src/memory/memoryManager.ts [hunk 4]
```diff
@@ -17,18 +18,38 @@ export class MemorySegmentManager {
     return ptr;
   }
 
-  loadData(address: Relocatable, data: MaybeRelocatable[]): Relocatable {
+  loadData(
+    address: Relocatable,
+    data: MaybeRelocatable[]
+  ): Result<Relocatable> {
     for (let index = 0; index < data.length; index++) {
```
[the remark this answers is not in the document]
Comment (greged93) [reply src/memory/memoryManager.ts:25]: mmmmm depends, isn't a `for` loop a little more readable?

File: src/vm/virtualMachine.ts (hunk 3 above)
Comment (greged93) [reply src/vm/virtualMachine.ts:52]: yes, currently working on the rest of the computation that will return the operands

File: src/memory/memoryManager.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Eikix) [reply src/memory/memoryManager.ts:25]: i feel as though forEach is slightly more readable imo

[the remark this answers is not in the document]
Comment (greged93) [reply src/memory/memoryManager.ts:25]: will switch to for then

File: src/result/error.ts (hunk 1 above)
Comment (greged93) [reply src/result/error.ts:16]: I had to do this `type ErrorTypeKey = (typeof ErrorType)[keyof typeof ErrorType];` becase you are actually passing the value of the ErrorType object.

File: src/result/memory.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+export const UnknownAddressError =
+  'Access memory at unknown or uninitialized address';
+export const WriteOnceError =
+  'Write existing memory. Can only write to memory once.';
+export const SegmentIncrementError = 'Error incrementing number of segments';
+export const InstructionError = 'Instruction must be a Field Element';
```
[the remark this answers is not in the document]
Comment (greged93) [reply src/result/memory.ts:6]: this is actually used in the virtual machine: we get the next instruction which can be found in the pc memory segment at the current pc offset. This returns a maybe relocatable. If this maybe relocatable is not a felt, we throw this error

**Review by Eikix — APPROVED** [review]

**Pull request merged**